src/uct/ib/rdmacm/rdmacm_iface.c

```diff
@@ -70,6 +71,225 @@ static uct_iface_ops_t uct_rdmacm_iface_ops = {
     .iface_get_address        = uct_rdmacm_iface_get_address
 };
 
+void uct_rdmacm_iface_client_start_next_ep(uct_rdmacm_iface_t *iface)
+{
+    size_t ip_len = ucs_max(INET_ADDRSTRLEN, INET6_ADDRSTRLEN);
+    char *ip_str  = ucs_alloca(ip_len);
+    unsigned set_new_ep = 0;
+
+    UCS_ASYNC_BLOCK(iface->super.worker->async);
+    /* set a new ep (take one from the pending eps queue) */
+    while (!ucs_queue_is_empty(&iface->pending_eps_q) && (!set_new_ep)) {
+        iface->ep = ucs_queue_pull_elem_non_empty(&iface->pending_eps_q,
+                                                  uct_rdmacm_ep_t, queue);
+
+        if (rdma_resolve_addr(iface->cm_id, NULL, iface->ep->remote_addr,
+                              UCS_MSEC_PER_SEC * iface->config.addr_resolve_timeout)) {
+            ucs_error("rdma_resolve_addr(addr=%s) failed: %m",
+                       ucs_sockaddr_str(iface->ep->remote_addr, ip_str, ip_len));
+            uct_rdmacm_ep_set_failed(&iface->super.super, &iface->ep->super.super,
+                                     UCS_ERR_IO_ERROR);
+        } else {
+            set_new_ep = 1;
+        }
+    }
+
+    if (!set_new_ep) {
+        iface->ep = NULL;
+    }
+
+    UCS_ASYNC_UNBLOCK(iface->super.worker->async);
+}
+
+static void uct_rdmacm_client_handle_failure(uct_rdmacm_iface_t *iface, ucs_status_t status)
+{
+    ucs_assert((!iface->is_server));
+    uct_rdmacm_ep_set_failed(&iface->super.super, &iface->ep->super.super, status);
+    uct_rdmacm_iface_client_start_next_ep(iface);
+}
+
+static void uct_rdmacm_iface_process_conn_req(uct_rdmacm_iface_t *iface,
+                                             struct rdma_cm_event *event,
+                                             struct sockaddr *remote_addr)
+{
+    uct_rdmacm_priv_data_hdr_t *hdr, cb_hdr;
+    size_t max_conn_priv;
+    ssize_t server_data_len;
+    struct rdma_conn_param conn_param;
+    ucs_status_t status;
+    size_t ip_len  = ucs_max(INET_ADDRSTRLEN, INET6_ADDRSTRLEN);
+    char *ip_str   = ucs_alloca(ip_len);
+
+    status = uct_base_iface_get_max_conn_priv(&iface->super.super, &max_conn_priv);
+    if (status != UCS_OK) {
+        ucs_error("Server failed to get max_conn_priv");
+        rdma_reject(event->id, NULL, 0);
+        return;
+    }
+
+    hdr = (uct_rdmacm_priv_data_hdr_t*) event->param.ud.private_data;
+
+    memset(&conn_param, 0, sizeof(conn_param));
+    conn_param.private_data = ucs_alloca(max_conn_priv + sizeof(uct_rdmacm_priv_data_hdr_t));
```

Comment:
@evgeny-leksikov I had _ucs_malloc_ there and changed it to _ucs_alloca_ since @yosefe suggested to use _ucs_alloca_ when allocating the same size on the client side. Also, this field is used only inside this function.

Reply:
max_conn_priv is pretty small (~136b) so alloca can be used for simplicity